include/onnxruntime/core/platform/EigenNonBlockingThreadPool.h

```diff
@@ -158,6 +164,73 @@ struct PaddingToAvoidFalseSharing {
   char padding[ORT_FALSE_SHARING_BYTES];
 };
 
+/* Usage:
+1. In executor, call Start() before profiling and Stop() to get profiled numbers;
+2. Inside thread pool, call LogStart() before interested section and LogEnd... after to log elapsed time;
+3. To extend, just add more events in enum Event before "All", and update GetEventName(...) accordingly;
+4. Note LogStart must pair with either LogEnd or LogEndAndStart, otherwise ORT_ENFORCE will fail;
+5. ThreadPoolProfiler is thread-safe.
+*/
+
+class ThreadPoolProfiler {
+ public:
+  enum ThreadPoolEvent {
+    DISTRIBUTION = 0,
+    DISTRIBUTION_ENQUEUE,
+    RUN,
+    WAIT,
+    WAIT_REVOKE,
+    MAX_EVENT
+  };
+  ThreadPoolProfiler(int num_threads, const CHAR_TYPE* threal_pool_name);
+  ~ThreadPoolProfiler();
+  ORT_DISALLOW_COPY_ASSIGNMENT_AND_MOVE(ThreadPoolProfiler);
+  using Clock = std::chrono::high_resolution_clock;
+  void Start();                  //called by executor to start profiling
+  std::string Stop();            //called by executor to stop profiling and return collected numbers
+  void LogStart();               //called in main thread to record the starting time point
+  void LogEnd(ThreadPoolEvent);  //called in main thread to calculate and save the time elapsed from last start point
+  void LogEndAndStart(ThreadPoolEvent);
+  void LogCoreAndBlock(std::ptrdiff_t block_size);  //called in main thread to log core and block size for task breakdown
+  void LogThreadId(int thread_idx);                 //called in child thread to log its id
```

Comment:
Should these be std::thread::id (as with thread_id_ below)?

Reply:
thread_idx here is just the index of the child thread in the pool, not the id of the thread to os.